Add NGINX Ingress controller with OSM tutorial (#3592)
nginxinc/kubernetes-ingress: docs/content/tutorials/nginx-ingress-osm.md

> ## This document outlines how to integrate F5 NGINX Ingress Controller with Open Service Mesh (OSM)

Reviewer: ```suggestion
---
title: Using NGINX Ingress Controller with Open Service Mesh
description: This document explains how to integrate F5 NGINX Ingress Controller and Open Service Mesh
weight: 1800
doctypes: ["concept"]
toc: true
docs: "DOCS-1181"
---

## This document outlines how to integrate F5 NGINX Ingress Controller with Open Service Mesh
```

Without the Hugo frontmatter, this document will not be displayed.

Reviewer 2: ```
---
title: NGINX Ingress Controller and Open Service Mesh
description: |
  Use NGINX Ingress Controller with Open Service Mesh.
weight: 1800
doctypes: ["concept"]
toc: true
docs: "DOCS-1181"
---
```

minor change to Alan's suggestion to keep in line with Istio page

Open Service Mesh will work with both versions of [F5 NGINX Ingress controller](https://github.com/nginxinc/kubernetes-ingress): the free as well as the NGINX Plus versions.

> Below is a link to the official F5 NGINX Ingress controller documentation.

Reviewer: ```suggestion
Below is a link to the official F5 NGINX Ingress Controller documentation.
```

[F5 NGINX Ingress controller](https://docs.nginx.com/nginx-ingress-controller/)

> # Integrating NGINX Ingress Controller with Open Service Mesh

Reviewer: ```suggestion
## Integrating NGINX Ingress Controller with Open Service Mesh
```

> 
> Below is a link to the official F5 NGINX Ingress controller documentation.
> [F5 NGINX Ingress controller](https://docs.nginx.com/nginx-ingress-controller/)
> 
> # Integrating NGINX Ingress Controller with Open Service Mesh

Reviewer: ```suggestion

## Integrating NGINX Ingress Controller with Open Service Mesh
```

There's no need to broadly link to the landing page because this document will exist within the docset: this means that someone can use the navigation to get to the rest of it.

h2 (##) is the highest level heading we typically use in docs. I will make this suggestion on every instance of a h1 (#) that I spot.

There are two ways to integrate the NGINX Ingress Controller with Open Service Mesh (OSM):

1. Injecting an envoy sidecar directly with NGINX Ingress Controller.
2. Using the Open Service Mesh `ingressBackend` "proxy" feature.

> # NGINX Ingress controller and OSM with sidecar injected

Reviewer: ```suggestion
## Integrating NGINX Ingress Controller and Open Service Mesh with sidecar injection
```

> Install OSM in the cluster

Reviewer: ```suggestion
To begin, install OSM to the cluster.
```

---
title: NGINX Ingress Controller and Open Service Mesh
description: |
  Use NGINX Ingress Controller with Open Service Mesh.
weight: 1800
doctypes: ["concept"]
toc: true
docs: "DOCS-1181"
---

## This document outlines how to integrate F5 NGINX Ingress Controller with Open Service Mesh (OSM)

Open Service Mesh will work with both versions of [F5 NGINX Ingress controller](https://github.com/nginxinc/kubernetes-ingress): the free as well as the NGINX Plus versions.

Below is a link to the official F5 NGINX Ingress Controller documentation.
[F5 NGINX Ingress controller](https://docs.nginx.com/nginx-ingress-controller/)

# Integrating NGINX Ingress Controller with Open Service Mesh

There are two ways to integrate the NGINX Ingress Controller with Open Service Mesh (OSM):

1. Injecting an envoy sidecar directly with NGINX Ingress Controller.
2. Using the Open Service Mesh `ingressBackend` "proxy" feature.

> # NGINX Ingress controller and OSM with sidecar injected

Reviewer 3: "NGINX Ingress controller with OSM sidecar proxy injected"

Install OSM in the cluster

```bash
osm install --mesh-name osm-nginx --osm-namespace osm-system
```

> ### Mark F5 NGINX Ingress controller namespace for sidecar injection

Reviewer 3: Mark **the** F5...

*NOTE:* Depending on how you install NGINX Ingress controller, you might need to create the `namespace`. For example, if you are using manifests to install NGINX Ingress controller, you can complete all of the steps on our documentation page, *EXCEPT*, actually deploying NGINX Ingress controller. This is because, when using the sidecar approach, OSM needs to "manage" the namespace so it knows what `namespaces` it needs to inject sidecars into.

Next thing we need to do is install OSM into the `NGINX Ingress controller` namespace so that the `envoy` sidecar will be injected into NGINX Ingress controller.
First, create the `nginx-ingress` namespace:

```bash
kubectl create ns nginx-ingress
```
Then "mark" the `nginx-ingress` namespace for OSM to deploy a sidecar.

```bash
osm namespace add nginx-ingress --mesh-name osm-nginx
```

The above command will use the mark the `nginx-ingress` namespace, where OSM will be installed (sidecar)

> # Install F5 NGINX Ingress controller. Here the links to the install guides:

Reviewer 3: Install F5 NGINX Ingress controller. Here the links to the install guides:
That ^ reads very awkwardly as a two-sentence headline.  Not to mention the second sentence is incomplete.

This could be simplified to:
# Install F5 NGINX Ingress controller. 

Links to the complete install guides:

[Using Helm to install NGINX Ingress](https://docs.nginx.com/nginx-ingress-controller/installation/installation-with-helm/)
[Using Manifests to install NGINX Ingress](https://docs.nginx.com/nginx-ingress-controller/installation/installation-with-manifests/)

> 

Reviewer 3: Do the customers need to link off and follow the full guides?
Or is the install also summarized below?

Author: This was requested by the OSM guys. To keep the docs as close together, I added it in our repo as well.

When using the sidecar method, ensure that you add the correct annotations listed below. This ensures proper integration of NGINX Ingress Controller with the envoy sidecar proxy.

## Helm installs

If using `helm`, add the following `annotation` to your `values.yaml` file:

Under `controller.pod.annotations`:

```yaml
pod:
  annotations: {
    openservicemesh.io/inbound-port-exclusion-list: "80, 443"
    }
```

You can also use the `set` command available with `helm` to set these at install time.

```bash
helm install nic01 nginx-stable/nginx-ingress -n nginx-ingress --create-namespace --set controller.pod.annotations.'openservicemesh\.io/inbound\-port\-exclusion\-list=\{ "80"\, "443"\ }'
```

Change your `release` accordingly to match your environment.

## Manifest installs

For your `manifest` deployments, add the following `annotation`.

```yaml
annotations:
  openservicemesh.io/inbound-port-exclusion-list: "80,443"
```

> ### Sample deployment file with annotation required

Reviewer 3: Sample deployment file with required annotation

```yaml
apiVersion: apps/v1
kind: Deployment
metadata:
  name: nginx-ingress
  namespace: nginx-ingress
spec:
  replicas: 1
  selector:
    matchLabels:
      app: nginx-ingress
  template:
    metadata:
      labels:
        app: nginx-ingress
      annotations:
        openservicemesh.io/inbound-port-exclusion-list: "80,443"
```

This annotation is *required* when injecting envoy sidecar into NGINX Ingress controller.
`InboundPortExclusionList` defines a global list of ports to exclude from inbound traffic interception by the sidecar proxy.

### Install a Test Application
To test the integration, we will use the `httpbin` sample application from the [Ingress With Kubernetes NGINX Ingress Controller](https://release-v1-2.docs.openservicemesh.io/docs/demos/ingress_k8s_nginx/) guide.

The following three commands will create the namespace for the application, add the namespace to OSM for monitoring, then install the application.

```bash
kubectl create ns httpbin
osm namespace add httpbin --mesh-name osm-nginx
kubectl apply -f https://raw.githubusercontent.com/openservicemesh/osm-docs/release-v1.2/manifests/samples/httpbin/httpbin.yaml -n httpbin
```

> ### Verify that the envoy sidecar has been *injected* into NGINX Ingress

Reviewer 3: NGINX Ingress Controller
(consistency)

```

Test your configuration:

```bash
 curl  http://httpbin.example.com/get -v
*   Trying 172.19.0.2:80...
* TCP_NODELAY set
* Connected to httpbin.example.com (172.19.0.2) port 80 (#0)
> GET /get HTTP/1.1
> Host: httpbin.example.com
> User-Agent: curl/7.68.0
> Accept: */*
>
* Mark bundle as not supporting multiuse
< HTTP/1.1 200 OK
< Server: nginx/1.23.3
< Date: Sun, 19 Feb 2023 19:06:47 GMT
< Content-Type: application/json
< Content-Length: 454
< Connection: keep-alive
< access-control-allow-origin: *
< access-control-allow-credentials: true
< x-envoy-upstream-service-time: 2
<
{
  "args": {},
  "headers": {
    "Accept": "*/*",
    "Host": "httpbin.httpbin.svc.cluster.local",
    "Osm-Stats-Kind": "Deployment",
    "Osm-Stats-Name": "httpbin",
    "Osm-Stats-Namespace": "httpbin",
    "Osm-Stats-Pod": "httpbin-78555f5c4b-t6qln",
    "User-Agent": "curl/7.68.0",
    "X-Envoy-Internal": "true",
    "X-Forwarded-Host": "httpbin.example.com"
  },
  "origin": "172.19.0.1",
  "url": "http://httpbin.example.com/get"
}
* Connection #0 to host httpbin.example.com left intact
```

## Using The Open Service Mesh `ingressBackend` "proxy" Feature

> Install OSM into cluster.

Reviewer 3: Install OSM into **the** cluster

By running the following command, you will install OSM into the cluster with the mesh name `osm-nginx` using the `osm-system` namespace.

```bash
osm install --mesh-name osm-nginx --osm-namespace osm-system
```

Once OSM has been installed, this next command will mark the NGINX Ingress Controller as part of the OSM mesh, while also disabling sidecar injection.

> *NOTE*: The nginx-ingress name can be created as part of the NGINX Ingress install process, or manually. It must be created before you "add" the namespace to nginx-ingress.

Reviewer 3: Should this read??:
The nginx-ingress name**space** can be created as part of the NGINX Ingress install process, or manually. If you are creating it manually it must be created before you "add" the namespace to nginx-ingress.

```bash
osm namespace add nginx-ingress --mesh-name osm-nginx --disable-sidecar-injection
```

> # You can now install NGINX Ingress Controller by one of th above methods listed: `helm` or `manifetsts`.

Reviewer 3: th**e**

Also, why is this instruction below what we have already linked them off to?
If they need to install we should be telling them that just before we provide the links.

Author: Can remove them. It is a duplicate basically.
